build_wheels.py

```diff
@@ -172,28 +177,239 @@ def assemble_requirements(idf_branches: List[str], idf_constraints: List[str]) -
+                    '!': '=',
+                    '==': '!='}
+    for old, new in replacements.items():
+        if old in specifier and '===' not in specifier:
+            new_spec = specifier.replace(old, new)
+    return new_spec
+
+
+def yaml_to_requirement(yaml_file:str, exclude: bool = False) -> set:
+    """Converts YAML defined package into packaging.requirements Requirement
+    which can be directly used with pip
+    - when exclude is set to True it makes opposite logic for Requirement to be excluded"""
+    with open(yaml_file, 'r') as f:
+        yaml_list = yaml.load(f, yaml.Loader)
+
+    requirements_set: set[Requirement] = set()
+
+    if not yaml_list:
+        return requirements_set
+
+    for package in yaml_list:
+        requirement_txt = f"{package['package_name']}"
+
+        if 'version' in package:
+            if not isinstance(package['version'], list):
+                requirement_txt += _change_specifier_logic(package['version']) if exclude else package['version']
+
+            if isinstance(package['version'], list):
+                version_list = (
+                    [f'{_change_specifier_logic(ver)}' if exclude else f'{ver}' for ver in package['version']]
+                    )
+
+                requirement_txt += ','.join(version_list)
+
+
+        if 'platform' in package and 'version' not in package:
+            if not isinstance(package['platform'], list):
+                requirement_txt += (
+                    f"; sys_platform != '{package['platform']}'" if exclude
+                    else f"; sys_platform == '{package['platform']}'"
+                    )
+
+            if isinstance(package['platform'], list):
+                platform_list = (
+                    [f"sys_platform != '{plf}'" if exclude
+                     else f"sys_platform == '{plf}'" for plf in package['platform']]
+                    )
+
+                requirement_txt += '; ' + ' or '.join(platform_list)
+
+
+        if 'platform' in package and 'version' in package:
+            requirement_old = f"{package['package_name']}"
+
+            if not isinstance(package['platform'], list):
+                requirement_txt += f"; sys_platform == '{package['platform']}'"
+                if exclude:
+                    requirement_old += f"; sys_platform != '{package['platform']}'"
+                    requirements_set.add(Requirement(requirement_old))
+
+            if isinstance(package['platform'], list):
+                platform_list = [f"sys_platform == '{plf}'" for plf in package['platform']]
+                requirement_txt += '; ' + ' or '.join(platform_list)
+
+                if exclude:
+                    platform_list_old =  [f"sys_platform != '{plf}'" for plf in package['platform']]
+                    requirement_old += '; ' + ' or '.join(platform_list_old)
+                    requirements_set.add(Requirement(requirement_old))
+
+
+        requirements_set.add(Requirement(requirement_txt))
```

Comment:
You do a lot of string concatenation in this function. You could just collect the strings into a list and join them at the end. 
```python
str_list.append('str1')
str_list.append('str2')
...
requirements_set.add(Requirement(''.join(str_list)))
```
Appending to string requires to allocate a new array, copy both strings there and de-allocate the original strings. Doing this iteratively is a big overhead. Python cannot optimize it.

The above will do concatenation only once (it can count the right array size in advance to fit there all the strings). Python can optimize it and could improve with future versions. Makes this sense to you?

Reply:
Yes, it makes. Thank you for explanation.